clicommand/git_credentials_helper_test.go

```diff
@@ -0,0 +1,83 @@
+package clicommand
+
+import (
+	"errors"
+	"strings"
+	"testing"
+)
+
+func TestParseGitCredentialInput(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name        string
+		lines       []string
+		expected    string
+		expectedErr error
+	}{
+		{
+			name: "happy path",
+			lines: []string{
+				"protocol=https",
+				"host=github.com",
+				"path=buildkite/agent",
+			},
+			expected: "https://github.com/buildkite/agent",
+		},
+		{
+			name: "missing protocol",
+			lines: []string{
+				"host=github.com",
+				"path=buildkite/agent",
+			},
+			expectedErr: errors.New("missing protocol in git credential input"),
+		},
+		{
+			name: "missing host",
+			lines: []string{
+				"protocol=https",
+				"path=buildkite/agent",
+			},
+			expectedErr: errors.New("missing host in git credential input"),
+		},
+		{
+			name: "missing path",
+			lines: []string{
+				"protocol=https",
+				"host=github.com",
+			},
+			expectedErr: errors.New("missing path in git credential input"),
+		},
+		{
+			name: "non-https protocol",
+			lines: []string{
+				"protocol=ssh",
+				"host=github.com",
+				"path=buildkite/agent",
+			},
+			expectedErr: errors.New("git remote must be using the https protocol to use Github App authentication"),
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			input := strings.Join(tc.lines, "\n")
+			actual, actualErr := parseGitURLFromCredentialInput(input)
+			if tc.expectedErr != nil {
+				if actualErr == nil {
+					t.Fatalf("expected parseGitURLFromCredentialInput(%q) = error(%q), got nil", input, tc.expectedErr)
+				}
+
+				if actualErr.Error() != tc.expectedErr.Error() {
+					t.Fatalf("expected parseGitURLFromCredentialInput(%q) = error(%q), got error(%q)", input, tc.expectedErr, actualErr)
+				}
+			}
+
+			if actual != tc.expected {
+				t.Fatalf("expected parseGitURLFromCredentialInput(%q) = %q got %q", input, tc.expected, actual)
```

Comment:
I have two main comments to add:

* Phrasing: I've seen "got vs want" wording and "actual vs expected" wording, but this mixes the two. I don't mind if you insist on "actual / expected" but be consistent. It's also jarring to me to read `functionCall(input) = expected` even if you prefix that with "expected" - to me it's more natural to say `functionCall(input) = got, want ...` (you called this function with this input, and it equalled this particular value)
* Comparing error strings (`actualErr.Error() != expectedErr.Error()`): The test will break immediately if the error message returned by the function changes. Consider returning a sentinel error value and comparing directly or with `errors.Is`. That also lets you remove the check against `nil`.

Reply:
oh, yeah, no i just completely fucked that up. the tests passed first time so i think i didn't need to put much thought into the failure messages 😅 will fix this up